larpandoracontent/LArMonitoring/EventClusterValidationAlgorithm.cc

```diff
@@ -151,181 +249,681 @@ void EventClusterValidationAlgorithm::GetHitParents(
+        const int parentPdg{std::abs(pParentMC->GetParticleId())};
+        if (parentPdg == PHOTON || parentPdg == E_MINUS)
+        {
+            if (parentPdg == E_MINUS)
+            {
+                hasAncestorElectron = true;
+            }
+            pCurrentMC = pParentMC;
+            continue;
+        }
+        pLeadingMC = pCurrentMC;
+        break;
+    }
+
+    // Don't fold "showers" that consist of only compton scatters
+    // Trying to prevents distant diffuse hits disconnected from a "real" bremm + pair production shower being clustered together
+    if (hasAncestorElectron || std::abs(pLeadingMC->GetParticleId()) == E_MINUS)
+    {
+        return pLeadingMC;
+    }
+    else if (this->CausesShower(pLeadingMC, 0))
+    {
+        return pLeadingMC;
+    }
+    else
+    {
+        return pMC;
+    }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool EventClusterValidationAlgorithm::CausesShower(const MCParticle *const pMC, int nDescendentElectrons) const
+{
+    if (nDescendentElectrons > 1)
+    {
+        return true;
+    }
+
+    if (std::abs(pMC->GetParticleId()) == E_MINUS)
+    {
+        nDescendentElectrons++; // Including the parent particle, ie. the first in the recursion, as a descendent
+    }
+    for (const MCParticle *pChildMC : pMC->GetDaughterList())
+    {
+        if (this->CausesShower(pChildMC, nDescendentElectrons))
+        {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+inline bool EventClusterValidationAlgorithm::IsEM(const pandora::MCParticle *const pMC) const
+{
+    const int pdg{std::abs(pMC->GetParticleId())};
+    return (pdg == E_MINUS || pdg == PHOTON);
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+const MCParticle* EventClusterValidationAlgorithm::FoldPotentialDeltaRayTo(const CaloHit *const pCaloHit, const MCParticle *const pMC) const
+{
+    // Not an electron -> not a delta ray -> do nothing
+    if (pMC->IsRootParticle() || pMC->GetParticleId() != E_MINUS)
+    {
+        return pMC;
+    }
+
+    // Did not come from a track-like particle -> not a delta ray -> do nothing
+    const MCParticle *const pParentMC{*(pMC->GetParentList().begin())};
+    const int parentPdg{std::abs(pParentMC->GetParticleId())};
+    if (parentPdg == PHOTON || parentPdg == E_MINUS || PdgTable::GetParticleCharge(parentPdg) == 0)
+    {
+        return pMC;
+    }
+
+    // Delta ray that does not start a shower and is short -> fold into parent particle
+    if (!this->CausesShower(pMC, 0) && (pMC->GetVertex() - pMC->GetEndpoint()).GetMagnitudeSquared() < m_deltaRayLengthThresholdSquared)
```

Comment:
As per earlier comment on wire pitch, at the moment I think this will fold back a lot of delta ray hits? ~2.5 cm before something is distinct?

Reply:
All of these are in mm, so the threshold is 0.25cm which I think is small enough to ignore